### DISCUSSION TITLE
Why `Transform` is used over Unreal's `FTransform`?

### DISCUSSION THREAD

**Discussion by realvictorprm** [nxrighthere/UnrealCLR#136] [Help] [answered]
Looking into the implementation of the recently added `AddInstances` I noticed that we have to copy the whole array because the Transforms need to be converted to be of type `FTransform`. Why aren't these used by default and would it be ok to add an additional API layer which makes use of these instead of `Transform`? I'm not a big fan of copying over 10k instances in some circumstances 😖

**Comment by nxrighthere** [comment, marked as answer]
It's necessary to use custom abstractions to avoid headaches with memory layout across different platforms between native and managed code. It also makes extending and upgrading easier without modifying the source code of the engine.

In general, it's hard to avoid memory copying in interoperability when you want to keep code safe and maintainable for the end-user.

**Reply by realvictorprm** [reply]
Thanks for the insight, that makes sense 🤔
